core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/SellerCreatesAndSignsTx.java

```diff
@@ -56,60 +56,54 @@ protected void run() {
+            Coin required = BsqSwapCalculation.getSellersBtcInputValue(trade, sellersTxSize, sellersTradeFee);
 
             // As fee calculation is not deterministic it could be that we toggle between a too small and too large
             // input. We would take the latest result before we break iteration. Worst case is that we under- or
             // overpay a bit. As fee rate is anyway an estimation we ignore that imperfection.
             while (iterations < 10 && !required.equals(previous)) {
-                tuple = tradeWalletService.getSellersBtcInputsForBsqSwapTx(required);
+                inputsAndChange = tradeWalletService.getSellersBtcInputsForBsqSwapTx(required);
                 previous = required;
 
                 // We calculate more exact tx size based on resulted inputs and change
-                sellersTxSize = BsqSwapCalculation.getTxSize(protocolModel.getTradeWalletService(), tuple.first, tuple.second.getValue());
-                required = BsqSwapCalculation.getSellersRequiredBtcInput(trade, sellersTradeFee, sellersTxSize);
+                sellersTxSize = BsqSwapCalculation.getTxSize(tradeWalletService, inputsAndChange.first, inputsAndChange.second.getValue());
+                required = BsqSwapCalculation.getSellersBtcInputValue(trade, sellersTxSize, sellersTradeFee);
+
                 iterations++;
             }
 
-            checkNotNull(tuple);
-            List<RawTransactionInput> sellersInputs = tuple.first;
-            long sellersChange = tuple.second.value;
-
-            protocolModel.setInputs(sellersInputs);
-            protocolModel.setChange(sellersChange);
-
-            long buyersBtcPayout = BsqSwapCalculation.getBuyersBtcPayoutAmount(sellersTradeFee,
-                    trade,
-                    getBuyersTxSize(),
-                    getBuyersTradeFee());
-            tradePeer.setPayout(buyersBtcPayout);
+            checkNotNull(inputsAndChange);
 
             List<RawTransactionInput> buyersBsqInputs = Objects.requireNonNull(tradePeer.getInputs());
-            List<RawTransactionInput> sellersBtcInputs = Objects.requireNonNull(sellersInputs);
 
-            Coin sellersBsqPayoutAmount = Coin.valueOf(protocolModel.getPayout());
-            String sellersBsqPayoutAddress = protocolModel.getBsqAddress();
+            List<RawTransactionInput> sellersBtcInputs = inputsAndChange.first;
+            protocolModel.setInputs(sellersBtcInputs);
+
+            Coin sellersBsqPayoutAmount = BsqSwapCalculation.getSellerBsqPayoutValue(trade, sellersTradeFee);
+            protocolModel.setPayout(sellersBsqPayoutAmount.getValue());
+            String sellersBsqPayoutAddress = protocolModel.getBsqWalletService().getUnusedAddress().toString();
+            protocolModel.setBsqAddress(sellersBsqPayoutAddress);
 
             Coin buyersBsqChangeAmount = Coin.valueOf(tradePeer.getChange());
```

Comment:
Would be safer to calculate this ourselves rather than taking from peer. Could verify that peer has the same value, but should not be necessary.

Reply:
You refer to tradePeer.getChange? Will be handled in process msg task as discussed above.